sopel/modules/ip.py

```diff
@@ -10,52 +10,242 @@
 def configure(config):
-    """
-    | name | example | purpose |
-    | ---- | ------- | ------- |
-    | GeoIP\\_db\\_path | /home/sopel/GeoIP/ | Path to the GeoIP database files |
-    """
     config.define_section('ip', GeoipSection)
     config.ip.configure_setting('GeoIP_db_path',
                                 'Path of the GeoIP db files')
-
+    config.ip.configure_setting('IPQS_key',
+                                'Access key for IPQS service')
+    config.ip.configure_setting('warn_threshold',
+                                'Addresses with scores >= this will generate an alert')
+    config.ip.configure_setting('malicious_threshold',
+                                'Addresses with scores >= this will be z-lined')
+    config.ip.configure_setting('warn_chans',
+                                'List of channels to warn when a suspicious user is detected. '
+                                'May be empty.')
+    config.ip.configure_setting('protect_chans',
+                                'List of channels to +R after malicious attempt to reg. '
+                                'May be empty.')
+
+safe_mode = True
 
 def setup(bot):
     bot.config.define_section('ip', GeoipSection)
 
+def alert(bot, alert_msg: str, log_err = False):
+    for channel in bot.config.ip.warn_chans:
+        bot.say(alert_msg, channel)
+    if log_err:
+        LOGGER.error(alert_msg)
+
+def get_exemption(host):
+    ip = None
+    if isinstance(host, ipaddress.IPv4Address) or isinstance(host, ipaddress.IPv6Address):
+        ip = host
+    else:
+        try:
+            ip = ipaddress.ip_address(socket.getaddrinfo(host, None)[0][4][0])
+        except:
+            raise
+    if not ip.is_global:
+        LOGGER.warn(f"Non-global IP {ip} seen.")
+        return "Non-global IP; internal network, localhost, etc."
+    for network, reason in exemptions.items():
+        if ip in network:
+            return reason
+    return None
+
+def fetch_IPQS_ip_score(
+    ip_addr: typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address],
+    key: str,
+    allow_public_access_points: bool = True,
+    lighter_penalties: bool = true,
+    strictness: int = 2,
+    fast: bool = False,
+    mobile: bool = False
+    ) -> tuple[float, bool, bool]: #score, is proxy, has recent abuse flag set
+    '''Perform lookup on a specific IP adress using ipqualityscore.com'''
+    params = urllib.parse.urlencode({
+        'allow_public_access_points': "true" if allow_public_access_points else "false",
+        'strictness': int(strictness),
+        # lowercase + handle None and other non-bool garbage
+        'fast': "true" if fast else "false",
+        'mobile': "true" if mobile else "false",
+        'lighter_penalties': "true" if lighter_penalties else "false",
+        })
+    # ip_addr sourced from server, not user, so sanitization already done
+    with urllib.request.urlopen(
+        f"https://ipqualityscore.com/api/json/ip/{key}/{str(ip_addr)}?{params}") as url:
```

Comment:
Since you have this in two places, it might make sense to have an API module that wraps requests? Mostly a nitpick though